File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/StreamingFeature.swift
// Streaming result support. For list-style operations it attaches a
// `result.stream` function yielding items so callers can consume them
// incrementally with `for item in result.stream!()` instead of materialising
// the whole list at once. A `chunkSize` groups items into `.list` batches
// when set; a `chunkDelay` (ms) paces delivery via the injectable `sleep`
// for offline tests.

import Foundation

public final class StreamingFeature: BaseFeature {
  private var client: ProjectNameSDK?
  private var options: VMap?

  // Activity tracking (mirrors the ts client._streaming record).
  public var opened = 0

  public override init() {
    super.init()
    version = "0.0.1"
    name = "streaming"
    active = true
  }

  public override func initFeature(_ ctx: Context, _ options: VMap) {
    client = ctx.client
    self.options = options
    active = foptBool(options, "active", false)
  }

  public override func preResult(_ ctx: Context) {
    if !active || !streamable(ctx) {
      return
    }
    guard let result = ctx.result else {
      return
    }

    result.streaming = true
    result.stream = { self.iterate(result) }

    opened += 1
  }

  private func iterate(_ result: Result) -> [Value] {
    let chunkDelay = foptInt(options, "chunkDelay", 0)
    let chunkSize = foptInt(options, "chunkSize", 0)
    let sleep = foptSleep(options)

    // Read lazily at stream() call time so downstream result processing
    // is reflected.
    let items = result.resdata.asList?.items ?? []

    var out: [Value] = []

    if chunkSize > 0 {
      var i = 0
      while i < items.count {
        if chunkDelay > 0 {
          sleep(chunkDelay)
        }
        let end = min(i + chunkSize, items.count)
        out.append(.list(Array(items[i..<end])))
        i += chunkSize
      }
      return out
    }

    for item in items {
      if chunkDelay > 0 {
        sleep(chunkDelay)
      }
      out.append(item)
    }
    return out
  }

  private func streamable(_ ctx: Context) -> Bool {
    let opname = ctx.op?.name ?? ""
    let ops = foptStrList(options, "ops") ?? ["list"]
    return ops.contains(opname)
  }
}

File: ts/project/.sdk/tm/kotlin/feature/StreamingFeature.kt
package KOTLINPACKAGE.feature

import java.util.NoSuchElementException
import java.util.function.IntConsumer
import java.util.function.Supplier

import KOTLINPACKAGE.core.Context
import KOTLINPACKAGE.core.Result
import KOTLINPACKAGE.core.SdkClient

// Streaming result support. For list-style operations it attaches a
// `result.stream` supplier yielding an iterator so callers can consume items
// incrementally. A `chunkSize` groups items into List batches; a `chunkDelay`
// (ms) paces delivery via the injectable `sleep`.
@Suppress("UNCHECKED_CAST")
class StreamingFeature : BaseFeature("streaming", "0.0.1", true) {

  private var client: SdkClient? = null
  private var options: MutableMap<String, Any?>? = null

  // Activity tracking (mirrors the ts client._streaming record).
  var opened = 0

  override fun init(ctx: Context, options: MutableMap<String, Any?>) {
    this.client = ctx.client
    this.options = options
    this.active = FeatureOptions.foptBool(options, "active", false)
  }

  override fun preResult(ctx: Context) {
    if (!this.active || !streamable(ctx)) {
      return
    }
    val result = ctx.result ?: return

    result.streaming = true
    result.stream = Supplier { iterate(result) }

    this.opened++
  }

  private fun iterate(result: Result): Iterator<Any?> {
    val chunkDelay = FeatureOptions.foptInt(this.options, "chunkDelay", 0)
    val chunkSize = FeatureOptions.foptInt(this.options, "chunkSize", 0)
    val sleep: IntConsumer = FeatureOptions.foptSleep(this.options)

    // Read lazily at stream() call time so downstream result processing
    // is reflected.
    val rd = result.resdata
    val items: List<Any?> = if (rd is List<*>) rd as List<Any?> else mutableListOf()

    return object : Iterator<Any?> {
      private var index = 0

      override fun hasNext(): Boolean {
        return index < items.size
      }

      override fun next(): Any? {
        if (!hasNext()) {
          throw NoSuchElementException()
        }
        if (chunkDelay > 0) {
          sleep.accept(chunkDelay)
        }
        if (chunkSize > 0) {
          val end = minOf(index + chunkSize, items.size)
          val batch = ArrayList(items.subList(index, end))
          index = end
          return batch
        }
        return items[index++]
      }
    }
  }

  private fun streamable(ctx: Context): Boolean {
    val opname = ctx.op.name
    var ops: List<String>? = FeatureOptions.foptStrList(this.options, "ops")
    if (ops == null) {
      ops = listOf("list")
    }
    for (o in ops) {
      if (o == opname) {
        return true
      }
    }
    return false
  }
}
